<a href="https://colab.research.google.com/github/ericwarren9/ST-590/blob/main/Warren_HW3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ST 590 Homework 3 By: Eric Warren

## Reading and Manipulating Data with `pandas`

### Part 1

 This question doesn't involve any programming. Suppose you have a pandas `DataFrame`. In a markdown cell, describe what each of the following operators/methods allows you to subset from a data frame (that is, describe all of the types of subsetting you can do with the operator):
- []
- .iloc[]
- .loc[]

- By using the [] method, we can return **either** row(s) **or** column(s). However, we cannot return both at the same time. To get the column(s) we need to do something like `df["column_name"]` or to get multiple we do `df[["column_name1", "column_name2", ...]]`. To get the row(s) we do it via slicing using the integer of what row number(s) we want. We must have at least more than one value present meaning that if we want the first row we cannot do `df[0]` and must do `df[0:1]`.
- By using the .iloc[] method, we can return **both** rows **and** columns. The rows and columns specified must be in integer format and the format of how we would subset via this method is `df.iloc[rows, columns]`. An example to get the first two rows and two columns would be `df.iloc[0:2, 0:2]`.
- By using the .loc[] method, we can return **both** rows **and** columns. The rows and columns specified can be in integer **or** name (or label) format. Note that subsetting in .loc[] includes the last value in slicing, which is different than the others. An example of we could use .loc[] is `df.loc[0:5, "column_name"]` which gives us the first **six** rows and the column "column_name".
- Note we can use all of these to subset with booleans using logicals.

### Part 2

Read in the BreastCancer.dat data file available in the assignment link. (Open the file in a program such as notepad or wordpad to determine the delimiter - although a program like notepad++ is a better choice.) Upload this file to your Colab notebook and read it in using a relative file path (just the file name). Save the data as an object called cancer_data.

In [4]:
import pandas as pd # Import pandas

cancer_data = pd.read_csv("BreastCancer.dat", sep = "\t") # Read in the data

Use the `.head()` method to look at the data.

In [3]:
cancer_data.head()

id  age            meno  size  grade  gradd1  gradd2  nodes    enodes  \
0   132   49   premenopausal    18      2       1       0      2  0.786628   
1  1575   55  Postmenopausal    20      3       1       1     16  0.146607   
2  1140   56  Postmenopausal    40      3       1       1      3  0.697676   
3   769   45   premenopausal    25      3       1       1      1  0.886921   
4   130   65  Postmenopausal    30      2       1       0      5  0.548812   

   pgr  er         hormon  rectime  censrec  _st  _d        _t  _t0  
0    0   0   no tamoxifen     1838        0    1   0  5.032170    0  
1    0   0   no tamoxifen      403        1    1   1  1.103354    0  
2    0   0   no tamoxifen     1603        0    1   0  4.388775    0  
3    0   4   no tamoxifen      177        0    1   0  0.484600    0  
4    0  36  had tamoxifen     1855        0    1   0  5.078713    0

Return just the `grade` column using the column attribute

In [5]:
cancer_data.grade

0      2
1      3
2      3
3      3
4      2
      ..
681    3
682    2
683    3
684    2
685    2
Name: grade, Length: 686, dtype: int64

Use the `.loc[]` method to print out all rows where the size is larger than 30.

In [6]:
cancer_data.loc[cancer_data['size'] > 30]

id  age            meno  size  grade  gradd1  gradd2  nodes    enodes  \
2    1140   56  Postmenopausal    40      3       1       1      3  0.697676   
5    1642   48   premenopausal    52      2       1       0     11  0.267135   
14    820   32   premenopausal    57      3       1       1     24  0.056135   
19   1185   58  Postmenopausal    45      3       1       1      4  0.618783   
30    777   49   premenopausal    55      2       1       0      7  0.431710   
..    ...  ...             ...   ...    ...     ...     ...    ...       ...   
672   157   59  Postmenopausal    45      2       1       0      6  0.486752   
673  1137   47   premenopausal    70      2       1       0      5  0.548812   
674  1171   48  Postmenopausal    35      1       0       0      2  0.786628   
675  1178   64  Postmenopausal    35      1       0       0      4  0.618783   
683  1525   57  Postmenopausal    35      3       1       1      1  0.886921   

      pgr   er         hormon  rectime  censrec  _st  _d        _t  _t0  
2       0    0   no tamoxifen     1603        0    1   0  4.388775    0  
5       0    0   no tamoxifen      842        1    1   1  2.305270    0  
14      0   13   no tamoxifen      448        1    1   1  1.226557    0  
19      0    0   no tamoxifen      370        1    1   1  1.013005    0  
30      0    0   no tamoxifen      679        1    1   1  1.859001    0  
..    ...  ...            ...      ...      ...  ...  ..       ...  ...  
672   739  526   no tamoxifen      740        0    1   0  2.026010    0  
673   796   24   no tamoxifen     1703        0    1   0  4.662560    0  
674   845  105  had tamoxifen     1578        0    1   0  4.320328    0  
675   858   15   no tamoxifen      722        0    1   0  1.976728    0  
683  1490  209  had tamoxifen     1342        0    1   0  3.674196    0  

[219 rows x 18 columns]

Use the `.loc[]` method to print out all rows where the size is greater than 30 and the grade is 3.

In [7]:
cancer_data.loc[(cancer_data['size'] > 30) & (cancer_data['grade'] == 3)]

id  age            meno  size  grade  gradd1  gradd2  nodes    enodes  \
2    1140   56  Postmenopausal    40      3       1       1      3  0.697676   
14    820   32   premenopausal    57      3       1       1     24  0.056135   
19   1185   58  Postmenopausal    45      3       1       1      4  0.618783   
32    588   56  Postmenopausal    40      3       1       1      1  0.886921   
37    735   48   premenopausal    45      3       1       1      1  0.886921   
..    ...  ...             ...   ...    ...     ...     ...    ...       ...   
597  1239   37   premenopausal    35      3       1       1      1  0.886921   
637   884   53   premenopausal    75      3       1       1     19  0.102284   
652  1488   66  Postmenopausal    42      3       1       1     11  0.267135   
664   346   63  Postmenopausal    45      3       1       1      2  0.786628   
683  1525   57  Postmenopausal    35      3       1       1      1  0.886921   

      pgr   er         hormon  rectime  censrec  _st  _d        _t  _t0  
2       0    0   no tamoxifen     1603        0    1   0  4.388775    0  
14      0   13   no tamoxifen      448        1    1   1  1.226557    0  
19      0    0   no tamoxifen      370        1    1   1  1.013005    0  
32      0    3   no tamoxifen     1013        0    1   0  2.773443    0  
37      0    0   no tamoxifen     1862        0    1   0  5.097878    0  
..    ...  ...            ...      ...      ...  ...  ..       ...  ...  
597   242   67   no tamoxifen      205        1    1   1  0.561259    0  
637   375  107   no tamoxifen      600        1    1   1  1.642710    0  
652   412  339  had tamoxifen      394        1    1   1  1.078713    0  
664   530  328   no tamoxifen     2148        0    1   0  5.880903    0  
683  1490  209  had tamoxifen     1342        0    1   0  3.674196    0  

[62 rows x 18 columns]

Use [] to return just the age, size, and grade columns.

In [8]:
cancer_data[['age', 'size', 'grade']]

age  size  grade
0     49    18      2
1     55    20      3
2     56    40      3
3     45    25      3
4     65    30      2
..   ...   ...    ...
681   51    30      3
682   64    26      2
683   57    35      3
684   44    21      2
685   80     7      2

[686 rows x 3 columns]

Use `.loc[]` to return the rows where meno is equal to premenopausal along with the age, size,
and grade columns.

In [10]:
# Assuming we do not need to return the meno column
cancer_data.loc[cancer_data['meno'] == 'premenopausal',
                cancer_data.columns.isin(['age', 'size', 'grade'])]

age  size  grade
0     49    18      2
3     45    25      3
5     48    52      2
6     48    21      3
7     37    20      2
..   ...   ...    ...
669   38    61      2
673   47    70      2
679   43    20      2
681   51    30      3
684   44    21      2

[290 rows x 3 columns]

### Part 3

There are two files about mosquitos available at:
- [https://www4.stat.ncsu.edu/~online/datasets/mosquito.txt](https://www4.stat.ncsu.edu/~online/datasets/mosquito.txt)
- [https://www4.stat.ncsu.edu/~online/datasets/mosquito2.txt](https://www4.stat.ncsu.edu/~online/datasets/mosquito2.txt)

Determine the delimiter and read in the mosquito.txt file as an object called `mosq_data`. We know from the file the delimiter is "&"

In [27]:
# Read in the first data set
mosq_data = pd.read_csv("https://www4.stat.ncsu.edu/~online/datasets/mosquito.txt",
                        sep = "&")

# Show that the data set read in
mosq_data

Day  Cage trt  Response
0     1     1  T2        42
1     2     1  T1        44
2     2     2  T2        42
3     3     1  T1        25
4     3     2  T3        33
5     3     3  T2        34
6     4     1  T1        38
7     5     2  T1        17
8     5     3  T2        25
9     6     1  T2        30
10    6     2  T3        24
11    6     3  T1        36
12    7     1  T2        37
13    7     2  T3        31
14    7     3  T1        23
15    8     1  T1        31
16    8     2  T2        30
17    9     1  T1        25
18    9     2  T2        37
19    9     3  T3        30
20   10     1  T2        30
21   10     2  T3        23
22   10     3  T1        25
23   11     1  T3        21
24   11     2  T1        22
25   11     3  T2        34
26   12     1  T2        37
27   12     2  T3        35
28   12     3  T1        39
29   13     1  T1        31
30   13     3  T2        23
31   14     1  T2        31
32   14     2  T1        32
33   14     3  T3        26
34   15     1  T3        31
35   15     2  T1        27
36   15     3  T2        41

Similarly, read in the mosquito2.txt file. Note this file doesn't contain column names! The columns are the same as the other file though. Use an attribute from `mosq_data` to assign the column names as you read in the data. Save this data as an object called `mosq_data2`. Note the delimiter here is tabs.

In [28]:
# Read in the second data set
mosq_data2 = pd.read_csv("https://www4.stat.ncsu.edu/~online/datasets/mosquito2.txt",
                        sep = "\t",
                        names = mosq_data.columns)

# Show that the data set read in
mosq_data2

Day  Cage trt  Response
0    16     1  T2        40
1    16     2  T3        37
2    17     1  T2        34
3    17     2  T3        39
4    18     1  T2        42
5    18     2  T3        41
6    19     1  T3        42
7    19     2  T3        39
8    20     1  T1        41
9    20     2  T2        32
10   21     1  T2        41
11   21     2  T3        39
12   22     1  T3        40
13   22     2  T2        37
14   23     1  T1        40
15   23     2  T3        37
16   24     1  T2        36
17   24     2  T2        41
18   25     1  T2        38
19   25     2  T1        34
20   26     1  T3        29
21   26     2  T1        37
22   27     1  T3        39
23   27     2  T1        36
24   28     1  T1        39
25   28     2  T3        39
26   29     1  T2        33
27   29     2  T3        39
28   30     1  T1        42
29   30     2  T2        42

Combine the two datasets into one data frame using the `concat()` function from `pandas`. We are going to name this `combined_data`.

In [29]:
# Combine the data
combined_data = pd.concat([mosq_data, mosq_data2])

# Show the concated data
combined_data

Day  Cage trt  Response
0     1     1  T2        42
1     2     1  T1        44
2     2     2  T2        42
3     3     1  T1        25
4     3     2  T3        33
..  ...   ...  ..       ...
25   28     2  T3        39
26   29     1  T2        33
27   29     2  T3        39
28   30     1  T1        42
29   30     2  T2        42

[67 rows x 4 columns]

## Summarizing Data Numerically

For this part, we'll use the StudentData.txt data that comes from the UCI machine learning repository.

### Read in the data

Read this dataset from the URL. Check out the first
few observations of the data. We will save our data as `student_data_math` since we are only looking at the math scores. The delimiter is ";".

In [32]:
student_data_math = pd.read_csv("student-mat.csv", sep = ";") # Read in data

student_data_math.head() # Show first couple of observations

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

### Summarize the Data

This data has many categorical variables and a few numeric. We are going to analyze it in the following subsections.

#### Categorical Variables

Create a one-way contingency table, a two-way contingency table, and a three-way contingency table. Interpret a number from each resulting table (that is, pick out a value produced and explain what that value means.)

First we will look at the one-way table. We are going to look at the number of each sex of student.

In [33]:
student_data_math.sex.value_counts()

F    208
M    187
Name: sex, dtype: int64

From here, we can see that we have slightly more females with them outnumbering the number of male students by 208 to 187.

Now we are going to look at a two way contingency table looking at the sex of a student and their family size.

In [34]:
pd.crosstab(student_data_math.sex,
            student_data_math.famsize)

famsize  GT3  LE3
sex              
F        156   52
M        125   62

There are a couple things to note. The first is that most students have a family size greater than 3, meaning that on top of their parents they have at least one other sibling. The second thing to note is that it seems the proportion of male students might be only children (or have family sizes less than or equal to 3) than their female counterparts.

Now we are going to look at a three way contingency table looking at the sex of a student, their family size, and school.

In [36]:
pd.crosstab([student_data_math.sex, student_data_math.famsize],
            student_data_math.school)

school        GP  MS
sex famsize         
F   GT3      139  17
    LE3       44   8
M   GT3      113  12
    LE3       53   9

On top of what we have seen before, we can clearly see that more math students go to Gabriel Pereira than Mousinho da Silveira being the other school. It seems like the proportion of male students to female students seem to be about the same at each school and same with the family sizes.

Create a conditional two-way table. That is, condition on one variable's setting and create a two-way
table.

First, by subsetting the data (say with .loc) and then creating the two-way table

In [41]:
# Subset the math students by only looking at GP school students
math_data_GP = student_data_math.loc[student_data_math['school'] == "GP"]

# Make two way table
pd.crosstab(math_data_GP.sex,
            math_data_GP.famsize)

famsize  GT3  LE3
sex              
F        139   44
M        113   53

Next, by creating a three-way table and subsetting it (in our case by school "GP")

In [42]:
# Save three way table from before
my_three_way_table = pd.crosstab([student_data_math.sex,
                                  student_data_math.famsize],
                                 student_data_math.school)

# Now make a two way table by subsetting to only show females
my_three_way_table["GP"]

sex  famsize
F    GT3        139
     LE3         44
M    GT3        113
     LE3         53
Name: GP, dtype: int64

#### Numeric variables (and across groups)

The numeric variables are age, absences, and the three test grades variables (G1, G2, and G3).

Find measures of center and spread for three of these variables (including G3 as one of them). Here we are going to look at centers (mean and median) and spread (being standard deviation) of three of the numeric variables being age, G1, and G3.

In [54]:
student_data_math[['age', 'G1', 'G3']].agg(["mean", "median", "std"])

age         G1         G3
mean    16.696203  10.908861  10.415190
median  17.000000  11.000000  11.000000
std      1.276043   3.319195   4.581443

Now we are going to repeat while subsetting the data by some grouping variable (in this case we will look at the just female students)

In [58]:
student_data_math[['age', 'G1', 'G3']] \
.loc[student_data_math['sex'] == "F"] \
.agg(["mean", "median", "std"])

age         G1         G3
mean    16.730769  10.620192   9.966346
median  17.000000  10.000000  10.000000
std      1.201696   3.232530   4.622338

Find measures of center and spread across a single grouping variable (in this case we will look at it by sex) for three of these variables (including G3 as one of them). Here we are going to look at centers (mean and median) and spread (being standard deviation) of three of the numeric variables being age, G1, and G3.

In [59]:
student_data_math.groupby('sex')[['age', 'G1', 'G3']].agg(["mean", "median", "std"])

age                          G1                          G3         \
          mean median       std       mean median       std       mean median   
sex                                                                             
F    16.730769   17.0  1.201696  10.620192   10.0  3.232530   9.966346   10.0   
M    16.657754   16.0  1.356181  11.229947   11.0  3.392839  10.914439   11.0   

               
          std  
sex            
F    4.622338  
M    4.495297

Here we can see how Male students might do a little better on their math scores despite seeming to be a little younger in age.

Find measures of center and spread across two grouping variables (sex and school) for three of these variables (including G3 as one of them). Here we are going to look at centers (mean and median) and spread (being standard deviation) of three of the numeric variables being age, G1, and G3.

In [61]:
student_data_math \
.groupby(['sex', 'school'])[['age', 'G1', 'G3']] \
.agg(["mean", "median", "std"])

age                          G1                          G3  \
                 mean median       std       mean median       std       mean   
sex school                                                                      
F   GP      16.579235   16.0  1.173426  10.579235   10.0  3.154102   9.972678   
    MS      17.840000   18.0  0.746101  10.920000   11.0  3.817940   9.920000   
M   GP      16.457831   16.0  1.263005  11.337349   11.0  3.458068  11.060241   
    MS      18.238095   18.0  0.995227  10.380952   10.0  2.747293   9.761905   

                             
           median       std  
sex school                   
F   GP       11.0  4.604622  
    MS       10.0  4.846992  
M   GP       11.0  4.594725  
    MS       10.0  3.491486

It is interesting to see that Male students at Gabriel Pereira seem to have higher test scores than other groups. It is also worth to note that the older students tend to go to Mousinho da Silveira.

Create a correlation matrix between all of the numeric variables

In [62]:
student_data_math.corr()

<ipython-input-62-a0f8682ae6ec>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  student_data_math.corr()


age      Medu      Fedu  traveltime  studytime  failures  \
age         1.000000 -0.163658 -0.163438    0.070641  -0.004140  0.243665   
Medu       -0.163658  1.000000  0.623455   -0.171639   0.064944 -0.236680   
Fedu       -0.163438  0.623455  1.000000   -0.158194  -0.009175 -0.250408   
traveltime  0.070641 -0.171639 -0.158194    1.000000  -0.100909  0.092239   
studytime  -0.004140  0.064944 -0.009175   -0.100909   1.000000 -0.173563   
failures    0.243665 -0.236680 -0.250408    0.092239  -0.173563  1.000000   
famrel      0.053940 -0.003914 -0.001370   -0.016808   0.039731 -0.044337   
freetime    0.016434  0.030891 -0.012846   -0.017025  -0.143198  0.091987   
goout       0.126964  0.064094  0.043105    0.028540  -0.063904  0.124561   
Dalc        0.131125  0.019834  0.002386    0.138325  -0.196019  0.136047   
Walc        0.117276 -0.047123 -0.012631    0.134116  -0.253785  0.141962   
health     -0.062187 -0.046878  0.014742    0.007501  -0.075616  0.065827   
absences    0.175230  0.100285  0.024473   -0.012944  -0.062700  0.063726   
G1         -0.064081  0.205341  0.190270   -0.093040   0.160612 -0.354718   
G2         -0.143474  0.215527  0.164893   -0.153198   0.135880 -0.355896   
G3         -0.161579  0.217147  0.152457   -0.117142   0.097820 -0.360415   

              famrel  freetime     goout      Dalc      Walc    health  \
age         0.053940  0.016434  0.126964  0.131125  0.117276 -0.062187   
Medu       -0.003914  0.030891  0.064094  0.019834 -0.047123 -0.046878   
Fedu       -0.001370 -0.012846  0.043105  0.002386 -0.012631  0.014742   
traveltime -0.016808 -0.017025  0.028540  0.138325  0.134116  0.007501   
studytime   0.039731 -0.143198 -0.063904 -0.196019 -0.253785 -0.075616   
failures   -0.044337  0.091987  0.124561  0.136047  0.141962  0.065827   
famrel      1.000000  0.150701  0.064568 -0.077594 -0.113397  0.094056   
freetime    0.150701  1.000000  0.285019  0.209001  0.147822  0.075733   
goout       0.064568  0.285019  1.000000  0.266994  0.420386 -0.009577   
Dalc       -0.077594  0.209001  0.266994  1.000000  0.647544  0.077180   
Walc       -0.113397  0.147822  0.420386  0.647544  1.000000  0.092476   
health      0.094056  0.075733 -0.009577  0.077180  0.092476  1.000000   
absences   -0.044354 -0.058078  0.044302  0.111908  0.136291 -0.029937   
G1          0.022168  0.012613 -0.149104 -0.094159 -0.126179 -0.073172   
G2         -0.018281 -0.013777 -0.162250 -0.064120 -0.084927 -0.097720   
G3          0.051363  0.011307 -0.132791 -0.054660 -0.051939 -0.061335   

            absences        G1        G2        G3  
age         0.175230 -0.064081 -0.143474 -0.161579  
Medu        0.100285  0.205341  0.215527  0.217147  
Fedu        0.024473  0.190270  0.164893  0.152457  
traveltime -0.012944 -0.093040 -0.153198 -0.117142  
studytime  -0.062700  0.160612  0.135880  0.097820  
failures    0.063726 -0.354718 -0.355896 -0.360415  
famrel     -0.044354  0.022168 -0.018281  0.051363  
freetime   -0.058078  0.012613 -0.013777  0.011307  
goout       0.044302 -0.149104 -0.162250 -0.132791  
Dalc        0.111908 -0.094159 -0.064120 -0.054660  
Walc        0.136291 -0.126179 -0.084927 -0.051939  
health     -0.029937 -0.073172 -0.097720 -0.061335  
absences    1.000000 -0.031003 -0.031777  0.034247  
G1         -0.031003  1.000000  0.852118  0.801468  
G2         -0.031777  0.852118  1.000000  0.904868  
G3          0.034247  0.801468  0.904868  1.000000

Now let us create the correlation matrix for just the variables we explored before. Those variables are age, absences, and the three test grades variables (G1, G2, and G3).

In [64]:
student_data_math[['age', 'absences', 'G1', 'G2', 'G3']].corr()

age  absences        G1        G2        G3
age       1.000000  0.175230 -0.064081 -0.143474 -0.161579
absences  0.175230  1.000000 -0.031003 -0.031777  0.034247
G1       -0.064081 -0.031003  1.000000  0.852118  0.801468
G2       -0.143474 -0.031777  0.852118  1.000000  0.904868
G3       -0.161579  0.034247  0.801468  0.904868  1.000000

We can see that there is high correlation between all the grade variables, but otherwise weak correlation seems to exist.